### PR TITLE
Feat: Support newer EAGLE-3 drafters

### PR DESCRIPTION
**Pull request by Dogacel**
## Motivation

We will have a paper-release soon and we would like our models to have day-0 support with the paper/model release. 

We will release checkpoints for gpt-oss-20b and gpt-oss-120b.

## Modifications

1) Multi-layer EAGLE-3: backwards compatible as we rename midlayer -> layers.0, this gives flexibility for users to run more models including vLLM supported ones.

2) Alternate norm positioning: this is about our upcoming paper release and new models.

3) Support EAGLE-3 backbones that take other than 3 aux hidden states: current models only use 3 hidden states but our models benefit from more hidden states.

## Accuracy Tests, Speed Tests and Profiling

Hardware GH200, VRAM 96 GB, vCPU: 64, RAM: 432 GiB hosted on lambda.ai

### MT-Bench Results (temperature=0.7, batch_size=1, depth=5, topk=1, max_tokens=6)

| Metric | Current (`zhuyksir/EAGLE3-gpt-oss-20b-bf16`) | New (ours) | Δ |
|---|---:|---:|---:|
| **Latency (s)** | 624.09 | **384.67** | −38.4% |
| **Throughput (tok/s)** | 224.53 | **365.15** | +62.6% |
| **Accept Length** | 1.547 | **2.646** | +71.1% |

### Per-Category Breakdown

| Category | Throughput (Old → New) | Δ | Accept Length (Old → New) | Δ |
|---|---:|---:|---:|---:|
| Writing | 163.58 → 292.93 | +79.1% | 1.483 → 2.534 | +70.9% |
| Roleplay | 258.29 → 390.37 | +51.1% | 1.504 → 2.502 | +66.4% |
| Reasoning | 173.00 → 281.20 | +62.5% | 1.557 → 2.593 | +66.5% |
| Math | 109.97 → 172.41 | +56.8% | 1.912 → **3.714** | **+94.2%** |
| Coding | 298.31 → 509.22 | +70.7% | 1.668 → 3.015 | +80.8% |
| Extraction | 134.42 → 263.02 | **+95.7%** | 1.570 → 2.999 | +91.0% |
| STEM | 321.15 → 494.25 | +53.9% | 1.486 → 2.468 | +66.1% |
| Humanities | 337.52 → 517.79 | +53.4% | 1.468 → 2.345 | +59.7% |

### vs. D-Flash

depth=5 seems to be the sweet-spot for GPT-oss for both models.

| Metric | d-flash | Ours | Δ |
|---|---:|---:|---:|
| **Latency (s)** | 501.69 | **384.67** | −23.3% |
| **Throughput (tok/s)** | 287.67 | **365.15** | +26.9% |
| **Accept Length** | 2.622 | **2.646** | +0.9% |

Accept length is essentially tied, but ours is meaningfully faster, ~27% higher throughput at the same draft quality.

### vs. Baseline (no speculative decoding, bs=1)

| Metric | Baseline | Ours | Δ |
|---|---:|---:|---:|
| **Latency (s)** | 441.63 | **384.67** | −12.9% |
| **Throughput (tok/s)** | 315.86 | **365.15** | +15.6% |

### Per-Category Throughput vs Baseline

| Category | Baseline → Ours | Δ |
|---|---:|---:|
| Writing | 221.94 → 292.93 | +32.0% |
| Roleplay | 370.14 → 390.37 | +5.5% |
| Reasoning | 233.55 → 281.20 | +20.4% |
| Math | 155.41 → 172.41 | +10.9% |
| Coding | 450.84 → 509.22 | +12.9% |
| Extraction | 180.30 → 263.02 | +45.9% |
| STEM | 445.21 → 494.25 | +11.0% |
| Humanities | 469.50 → 517.79 | +10.3% |

## Checklist

- [x] Format your code according to the [Format code with pre-commit](https://docs.sglang.io/developer_guide/contribution_guide.html#format-code-with-pre-commit).
- [x] Add unit tests according to the [Run and add unit tests](https://docs.sglang.io/developer_guide/contribution_guide.html#run-and-add-unit-tests).
- [x] Update documentation according to [Write documentations](https://docs.sglang.io/developer_guide/contribution_guide.html#write-documentations).
- [x] Provide accuracy and speed benchmark results according to [Test the accuracy](https://docs.sglang.io/developer_guide/contribution_guide.html#test-the-accuracy) and [Benchmark the speed](https://docs.sglang.io/developer_guide/contribution_guide.html#benchmark-the-speed).
- [x] Follow the SGLang code style [guidance](https://docs.sglang.io/developer_guide/contribution_guide.html#code-style-guidance).

## Review and Merge Process

1. Ping Merge Oncalls to start the process. See the [PR Merge Process](https://github.com/sgl-project/sglang/blob/main/.github/MAINTAINER.md#pull-request-merge-process).
2. Get approvals from [CODEOWNERS](https://github.com/sgl-project/sglang/blob/main/.github/CODEOWNERS) and other reviewers.
3. Trigger CI tests with [comments](https://docs.sglang.io/developer_guide/contribution_guide.html#how-to-trigger-ci-tests) or contact authorized users to do so.
   - Common commands include `/tag-and-rerun-ci`, `/tag-run-ci-label`, `/rerun-failed-ci`
4. After green CI and required approvals, ask Merge Oncalls or people with Write permission to merge the PR.

### CODE REVIEW

**Comment by Dogacel** [comment]
@Qiaolin-Yu 🙏

File: python/sglang/srt/models/llama_eagle3.py [hunk 1]
```diff
@@ -269,6 +302,16 @@ def load_weights(self, weights: Iterable[Tuple[str, torch.Tensor]]) -> None:
         ]
 
         for name, loaded_weight in weights:
+            if "midlayer" in name:
+                name = name.replace("midlayer", "layers.0")
```
Comment (Qiaolin-Yu) [inline python/sglang/srt/models/llama_eagle3.py]: Is this a little hacky?

File: python/sglang/srt/models/llama_eagle3.py [hunk 2]
```diff
@@ -183,34 +216,37 @@ def forward(
...
             hidden_states, residual
         )
 
-        # For draft decode, we capture the hidden state before norm
-        return hidden_states_to_logits, [hidden_states_to_aux]
+        # For draft decode, we capture the hidden state before norm, but some models might prefer normed hidden states
+        draft_decode = (
+            [hidden_states] if not self.norm_output else [hidden_states_to_logits]
+        )
+        return hidden_states_to_logits, draft_decode
```
Comment (Qiaolin-Yu) [inline python/sglang/srt/models/llama_eagle3.py]: why not use `hidden_states_to_aux` here?

File: python/sglang/srt/models/llama_eagle3.py (hunk 1 above)
Comment (Dogacel) [reply python/sglang/srt/models/llama_eagle3.py]: My goal was to make old checkpoints still compatible, since different checkpoints have different names.

I.e. Redhat's has layers.0: https://huggingface.co/RedHatAI/gpt-oss-20b-speculator.eagle3/blob/main/model.safetensors

LMsys' have midlayer.

File: python/sglang/srt/models/llama_eagle3.py (hunk 2 above)
Comment (Dogacel) [reply python/sglang/srt/models/llama_eagle3.py]: Makes sense 👍

File: python/sglang/srt/speculative/eagle_worker.py [hunk 3]
```diff
@@ -1131,13 +1131,27 @@ def forward_draft_extend_after_decode(
...
-                if self.speculative_algorithm.is_eagle3()
+            if (
+                self.speculative_algorithm.is_eagle3()
                 and self.eagle_use_aux_hidden_state
-                else target_cfg.spec_hidden_size
-            )
-            draft_extend_input = EagleDraftExtendInput.create_idle_input(
+            ):
+                eagle_config = getattr(target_cfg.hf_config, "eagle_config", None) or {}
+                num_aux = getattr(target_cfg.hf_config, "num_aux_hidden_states", None)
```
Comment (kpham-sgl) [inline python/sglang/srt/speculative/eagle_worker.py]: ditto

File: python/sglang/srt/speculative/eagle_worker.py [hunk 4]
```diff
@@ -1131,13 +1131,27 @@ def forward_draft_extend_after_decode(
...
+                    )
+                target_hidden = getattr(
+                    target_cfg.hf_config,
+                    "target_hidden_size",
+                    target_cfg.hidden_size,
+                )
+                hidden_size = target_hidden * num_aux
+            else:
+                hidden_size = target_cfg.spec_hidden_size
+            draft_extend_input = EagleDraftInput.create_idle_input(
```
Comment (kpham-sgl) [inline python/sglang/srt/speculative/eagle_worker.py]: Shouldn't this be `EagleDraftExtendInput`?

File: python/sglang/srt/speculative/multi_layer_eagle_worker.py [hunk 5]
```diff
@@ -677,15 +677,28 @@ def forward_draft_extend_after_decode(
...
+                    )
+                target_hidden = getattr(
+                    target_cfg.hf_config,
+                    "target_hidden_size",
+                    target_cfg.hidden_size,
+                )
+                hidden_size = target_hidden * num_aux
+            else:
+                hidden_size = target_cfg.hidden_size
+            draft_extend_input = EagleDraftInput.create_idle_input(
```
Comment (kpham-sgl) [inline python/sglang/srt/speculative/multi_layer_eagle_worker.py]: Ditto

File: python/sglang/srt/speculative/eagle_worker.py [hunk 6]
```diff
@@ -1131,13 +1131,27 @@ def forward_draft_extend_after_decode(
...
-            draft_extend_input = EagleDraftExtendInput.create_idle_input(
+            ):
+                eagle_config = getattr(target_cfg.hf_config, "eagle_config", None) or {}
+                num_aux = getattr(target_cfg.hf_config, "num_aux_hidden_states", None)
+                if num_aux is None:
+                    num_aux = len(
+                        eagle_config.get(
+                            "eagle_aux_hidden_state_layer_ids", [None, None, None]
+                        )
+                    )
```
Comment (kpham-sgl) [inline python/sglang/srt/speculative/eagle_worker.py]: https://github.com/sgl-project/sglang/pull/24826 recently introduced `_get_eagle_aux_layer_count`. Maybe we can refactor/reuse that here? cc @Qiaolin-Yu

File: python/sglang/srt/speculative/multi_layer_eagle_worker.py [hunk 7]
```diff
@@ -677,15 +677,28 @@ def forward_draft_extend_after_decode(
...
+            target_cfg = self.target_worker.model_runner.model_config
+            if self.speculative_algorithm.is_eagle3():
+                eagle_config = getattr(target_cfg.hf_config, "eagle_config", None) or {}
+                num_aux = getattr(target_cfg.hf_config, "num_aux_hidden_states", None)
+                if num_aux is None:
+                    num_aux = len(
+                        eagle_config.get(
+                            "eagle_aux_hidden_state_layer_ids", [None, None, None]
+                        )
+                    )
```
Comment (kpham-sgl) [inline python/sglang/srt/speculative/multi_layer_eagle_worker.py]: ditto

File: python/sglang/srt/speculative/eagle_worker.py (hunk 3 above)
Comment (kpham-sgl) [inline python/sglang/srt/speculative/eagle_worker.py]: A bit confused here, why are we reading from target config here for `num_aux`

**Review by pyc96 — APPROVED** [review]
LGTM regarding use_fc_norm related changes

**Review by hnyls2002 — APPROVED** [review]

**Comment by hnyls2002** [comment]
/rerun-test test_eagle_infer_a.py test_eagle_infer_beta.py test_eagle_dp_attention.py
